File: binary-tree-level-order-traversal/binary-tree-level-order-traversal.cc
#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
typedef TreeNode *node;
typedef vector<int> vi;
typedef vector<vi> vvi;
class Solution {
public:
  vvi levelOrder(node root) {
    vvi res;
    if(!root) return res;
    queue<node> Q[2];
    int sw = 0;
    Q[sw].push(root);
    vi level;
    while(!Q[0].empty() || !Q[1].empty()) {
      assert(!Q[sw].empty());
      node top = Q[sw].front();
      Q[sw].pop();
      level.push_back(top->val);
      if(top->left) Q[1-sw].push(top->left);
      if(top->right) Q[1-sw].push(top->right);
      if(Q[sw].empty()) {
        sw = 1-sw;
        res.push_back(level);
        level.clear();
      }
    }
    return res;
  }
};

const int N = 7;
int main(int argc, const char *argv[])
{
  int A[N+1] = 
  { 0,
    3,
    9,20,
    0,0,15,7,
  };
  node nodes[N+1] = { 0 };
  for(int i=1; i<N+1; i++) {
    if(A[i]) nodes[i] = new TreeNode(A[i]);
  }
  for(int i=1; i<N+1; i++) {
    if(2*i < N+1 && nodes[i]) nodes[i]->left = nodes[2*i];
    if(2*i+1 < N+1 && nodes[i]) nodes[i]->right = nodes[2*i+1];
  }
  puts("level order");
  int lv = 0;
  while((1<<lv) <= N) ++lv;
  for(int i=0; i<lv; i++) {
    for(int j=0; j<(1<<i) && (1<<i)+j < N+1; j++) 
      if(nodes[(1<<i)+j]) printf("%d ", nodes[(1<<i)+j]->val);
      else printf("# ");
    puts("");
  }
  puts("--------");

  Solution so;
  auto res = so.levelOrder(nodes[1]);
  cout << res.size() << endl;
  for(auto ii : res) {
    for(auto i : ii) printf("%d ", i);
    puts("");
  }
  return 0;
}

File: rs/src/quests/binary_tree_level_order_traversal.rs
/**
 * [102] Binary Tree Level Order Traversal
 *
 * Given the root of a binary tree, return the level order traversal of its nodes' values. (i.e., from left to right, level by level).
 
Example 1:
Input: root = [3,9,20,null,null,15,7]
Output: [[3],[9,20],[15,7]]

Example 2:
Input: root = [1]
Output: [[1]]

Example 3:
Input: root = []
Output: []

 
Constraints:

	The number of nodes in the tree is in the range [0, 2000].
	-1000 <= Node.val <= 1000

 */
pub struct Solution {}
use super::super::util::tree::{TreeNode, to_tree};

// submission codes start here

// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
use std::collections::VecDeque;
impl Solution {
    pub fn level_order(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<Vec<i32>> {
        let mut ans = vec![];
        if let Some(root) = root {
            let mut q = VecDeque::new();
            q.push_back(root);
            while !q.is_empty() {
                let mut n = q.len();
                let mut level = vec![];
                while let Some(top) = q.pop_front() {
                    level.push(top.borrow().val);
                    if let Some(x) = top.borrow_mut().left.take() { q.push_back(x); }
                    if let Some(x) = top.borrow_mut().right.take() { q.push_back(x); }
                    n -= 1;
                    if n == 0 { break; }
                }
                ans.push(level);
            }
        }
        ans
    }
}

// submission codes end

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() {
        println!("{:?}", Solution::level_order(tree![3,9,20,null,null,15,7]));
    }
}
